docs/ECN-test-plan.md

```diff
@@ -0,0 +1,145 @@
+
+
+  ![](image/RED_ECN_marking_algorithm.PNG)
+
+
+Commodity switches can run RED at ingress (enqueue packet to the switch buffer) or egress (dequeue packet from the switch buffer). Compared to ingress RED/ECN, egress RED/ECN can achieve lower feedback delay.
+
+### Scope
+
+The test cases depicted in this document aim to do functional testing of ECN behavior of SONiC DUT (Device Under Test) as per RED (Random Early Detection) algorithm.
+
+### Testbed
+
+```
++-------------+      +--------------+      +-------------+       
+| Keysight TX |------|   SONiC DUT  |------| Keysight RX | 
++-------------+      +--------------+      +-------------+ 
+
+Keysight ports are connected via SONiC switch as shown in the illustration above.
+```
+## Setup configuration
+
+### DUT Configuration
+•	PFC watchdog is disabled so that the SONiC DUT does not discard packets from the egress queue as  PFC PAUSE storm mitigation action.
+
+•	Enable ECN at queue 3.
+
+•	Configure minimum threshold to Kmin Bytes.
+
+•	Configure maximum threshold to Kmax Bytes.
+
+•	Configure marking probability as Pmax %.
+
+### Keysight configuration
+•	All Keysight ports should have the same bandwidth capacity.
+
+•	Test specific configurations are mentioned in respective test cases.
+
+## Test Cases
+
+### Test Case #1 - ECN marking at egress
+
+#### Test Objective
+
+This test aims to verify the DUT’s dequeue based ECN marking behavior (Egress marking).
+
+#### Test Configuration
+
+- On SONiC DUT configure the following:
+  1. A single lossless priority Pi (0 <= i <= 7).
+  2. Configure minimum and maximum ECN marking threshold of the lossless priority Pi to N KB (e.g., 100 KB).
+  3. DUT should have enough buffer to hold 2N KB packets.
+
+- Configure following traffic items on the Keysight device:
+  1. Test data traffic: A traffic item from the Keysight Tx port to
+        the Keysight Rx port. This traffic item should have DSCP value mapped to lossless priority Pi configured in the DUT. Number of packets should be fixed to 2N (e.g., 200), size of each packet should be 1KB each.
+  2. PFC PAUSE storm: Persistent PFC pause frames from the Keysight
+        Rx port to the Keysight Tx port. The priorities of PFC pause
+        frames should be same as that of 'Test data traffic'. And the
+        inter-frame transmission interval should be lesser than
+        per-frame pause duration.
+
+#### Test Steps
+
+1. Start PFC PAUSE storm to fully block the lossless priority at the
+    DUT.
+2. From the Keysight Tx port, send 2N data packets to the receiver. The packets should be mapped to priority Pi on the DUT.
+3. After all the test data packets are sent, start data capture on the Keysight Rx port. Then, stop PFC PAUSE.
```

Comment:
We should start packet capturing before sending test data packets. Otherwise, we may miss data packets. See https://github.com/baiwei0427/sonic-mgmt/blob/ixia/tests/ixia/EGRESS_ECN_README.md for the right order.

Reply:
Updated.